### PR TITLE
[MM-25767] Quick switch users search is always falling back to the database

### PR DESCRIPTION
**Pull request by ethervoid**
#### Summary

This PR fixes multiple problems we have when filtering the list of allowed channels in different situations: with or with team id, with view restrictions, etc

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-25767

### CODE REVIEW

**Comment by ethervoid** [comment]
@mgdelacroix and @jespino please take a look at this draft. I need to include more coverage and test different options (API, manually testing, etc) but I want to know your opinion because this is a sensitive change

File: api4/user_test.go [hunk 1]
```diff
@@ -997,6 +997,7 @@ func TestAutocompleteUsersInChannel(t *testing.T) {
 		Username        string
 		ExpectedResults int
 		MoreThan        bool
+		Exception       bool
```
Comment (jespino) [inline api4/user_test.go:1000]: `Exception`? maybe you mean `Error` or `ShouldFail` or something like that?

File: store/searchlayer/user_layer.go [hunk 2]
```diff
@@ -146,45 +145,48 @@ func (s *SearchUserStore) autocompleteUsersInChannelByEngine(engine searchengine
 	return autocomplete, nil
 }
 
+// We're contemplating different scenarios in this method:
```
Comment (jespino) [inline store/searchlayer/user_layer.go:148]: In godoc is a good practice to use the name of the function as the beginning of the explanation.

**Comment by ethervoid** [comment]
@jespino done! @mgdelacroix I've check if the guest user is only returning the channel they're in and is working as expected

File: store/searchlayer/user_layer.go [hunk 3]
```diff
@@ -146,45 +145,48 @@ func (s *SearchUserStore) autocompleteUsersInChannelByEngine(engine searchengine
 	return autocomplete, nil
 }
 
+// getListOfAllowedChannelsForTeam We're contemplating different scenarios in this method:
```
Comment (jespino) [inline store/searchlayer/user_layer.go:148]: actually is more a explanation, something like `getListOfAllowedChannelsForTeam return the list of allowed channels contemplating different scenarios:`

File: api4/user.go [hunk 4]
```diff
@@ -852,9 +852,18 @@ func autocompleteUsers(c *Context, w http.ResponseWriter, r *http.Request) {
 	}
 
 	if len(channelId) > 0 {
-		// Applying the provided teamId here is useful for DMs and GMs which don't belong
-		// to a team. Applying it when the channel does belong to a team makes less sense,
-		// but the permissions are checked above regardless.
+		// We're using the channelId to search for users inside that channel and the team
```
Comment (mgdelacroix) [inline api4/user.go:855]: :heart:

File: api4/user.go [hunk 5]
```diff
@@ -852,9 +852,18 @@ func autocompleteUsers(c *Context, w http.ResponseWriter, r *http.Request) {
...
+		// to get the not in channel list. Also we want to include the DM and GM users for
+		// that team which could only be obtained having the team id.
+		if len(teamId) == 0 {
+			c.Err = model.NewAppError("autocompleteUser",
+				"api.user.autocomplete_users.missing_team_id.app_error",
+				nil,
+				"channelId="+channelId+", teamId="+teamId,
```
Comment (mgdelacroix) [inline api4/user.go:862]: Having the teamId here seems unnecessary:

```suggestion
				"channelId="+channelId",
```

**Pull request merged**